### ISSUE TITLE
Bug: include and import directive "unexpected" errors

### ISSUE THREAD

**Issue by sloanlance** [stedolan/jq#1483]
When filters use module `include` or `import` directives in the middle, jq gives error messages like:

```txt
jq: error: syntax error, unexpected include (Unix shell quoting issues?) at <top-level>, line 1:
"fubar" | include "tarfu";.
jq: 1 compile error
```

However, the documentation doesn't state that module directives may appear _only_ at the beginning of filters, but that appears to be the case.

I've been using a JavaScript module ([sanack/node-jq](/sanack/node-jq)) to execute jq queries from a Node.js application.  Two of the arguments it accepts are the input and the jq filter.  I learned it works by concatenating the input and the filter together, with a vertical bar between them, into a single string.  Then it passes the string to jq on a command line, like this:

```txt
jq -n '<input> | <filter>'
```

I ran into trouble with this because my filter contains `include` directives.  For (a contrived) example, if my input were `"fubar"` and my filter were `include "tarfu";.`, the jq command would be:

```bash
jq -n '"fubar" | include "tarfu";.'
```

That caused jq to give the error message as shown at the top of this text.  However, if I manually change the filter, moving the `include` directive to its beginning, and run it on the command line myself, it works:

```bash
jq -n 'include "tarfu";"fubar" | .'
```

```txt
"fubar"
```

So, it seems that either:

1. jq has a bug that doesn't allow module directives in the middle of filters.  It should be fixed to allow `import` and `include` where appropriate.
1. Module directives are meant to appear only at the beginning of filters, but the documentation (1.5 and dev. manuals) don't state this.  The documentation should be specific about where these directives are allowed.

**Comment by wtlangford** [comment]
It's the latter. That's a documentation bug. Thanks for finding it!

I'll try to update the documentation soon (and do a million other things
I've been putting off)

**Comment by purpliminal** [comment]
This is definitely a documentation issue on jq's side, but I think it's also an actual bug report or possible improvement in node-jq? Instead of using `jq -n 'input | filter'`, they should probably pass the input via STDIN to `jq 'filter'`. If you could report the issue over to them, that would be great!

**Comment by sloanlance** [comment]
Thanks for the response, @wtlangford.  I'm happy to help, but honestly, I'm
a little disappointed that it's not a bug in jq.

Is there any reason `import` and `include` shouldn't be allowed in the
middle of filters?  I've not had any need for it before.  Aside from this
situation, I imagine it could be useful for conditional inclusions.

Should I open a feature request for `import` and `include` flexibility?

--
Lance E Sloan
"...just this guy, you know?"

**Comment by sloanlance** [comment]
Thanks for the response, @wtlangford.  I'm happy to help, but honestly, I'm
a little disappointed that it's not a bug in jq.

Is there any reason `import` and `include` shouldn't be allowed in the
middle of filters?  I've not had any need for it before.  Aside from this
situation, I imagine it could be useful for conditional inclusions.

Should I open a feature request for `import` and `include` flexibility?

**Comment by sloanlance** [comment]
Yes, @purpliminal, I had opened an issue for node-jq:  sanack/node-jq#104

I mentioned in the issue that node-jq's input to jq should be sent via `stdin` or a temporary file, but I've not yet received a response.  Depending on how the conversation goes, I might make a new feature request for it there.  In addition to resolving this problem, I suggested it because I thought there will be trouble later when I try to send a very large dataset through node-jq.  I expect it to exceed command line length limits.

This has made me think about better Node.js support for jq.  The two npm packages I've tried use somewhat kludgy techniques:

* ([sanack/node-jq](/sanack/node-jq)): Installs its own private copy of jq and invokes it via command line.  jq will run quickly, but there are potential I/O issues between it and node-jq.
* ([fiatjaf/jq-web](/fiatjaf/jq-web)): Uses Emscripten to make jq run in the JS environment.  It's slow, but more portable.  It also had a problem with `include` in my filter, but it may be for a different reason.  I'm not sure yet, but it may be filesystem access trouble.

I wonder why nobody's tried making a jq C++ Addon for Node.js.  It could make the entire jq API available to JavaScript programs.

Does the jq website have a list of ports to other languages or interfaces with them?  Maybe a curated list that highlights the best ones?

**Comment by pkoppstein** [comment]
> Does the jq website have a list of ports to other languages or interfaces with them?

See "Language Bindings" in the jq FAQ: https://github.com/stedolan/jq/wiki/FAQ#language-bindings

**Comment by trothwell** [comment]
Is there an example on how to structure modules?

**Comment by pkoppstein** [comment]
@trothwell - Does this help: https://github.com/stedolan/jq/wiki/Cookbook#include-or-import-a-module-and-call-its-functions
